src/qunit.cpp

```diff
@@ -348,11 +335,29 @@ bool QUnit::M(bitLenInt qubit)
 {
     bool result = shards[qubit].unit->M(shards[qubit].mapped);
 
-    /*
-     * Decomposes all of the bits in the shard, performing M() on each one and
-     * setting each new CU to the appropriate value.
-     */
-    Decompose(qubit);
+    QInterfacePtr unit = shards[qubit].unit;
+    bitLenInt mapped = shards[qubit].mapped;
+
+    if (unit->GetQubitCount() == 1) {
+        /* If we're keeping the bits, and they're already in their own unit, there's nothing to do. */
+        return result;
+    }
+
+    QInterfacePtr dest = CreateQuantumInterface(engine, engine, 1, 0, rand_generator);
+    unit->Decohere(mapped, 1, dest);
+
+    /* Update the mappings. */
+    for (auto &&shard : shards) {
+        if (shard.unit == unit && shard.mapped >= mapped) {
+            if (shard.mapped == mapped) {
+                shard.unit = dest;
+                shard.mapped -= mapped;
```

Comment:
For clarity, = 0.

Reply:
Actually, this whole branch can be replaced by (outside the loop) a `shards[qubit].unit = dest, mapped = 0`, which reduces the complexity a bit by eliminating the if() clause completely.